src/charm.py

```diff
@@ -0,0 +1,301 @@
+        self.manifest_file_root = "./src/files/manifests/"
+
+        self._lightkube_client = None
+
+    def _install(self, event):
+        self.logger.info("Installing by instantiating Kubernetes objects")
+        self.unit.status = MaintenanceStatus("Instantiating Kubernetes objects")
+
+        # TODO: catch error when this fails due to permissions and set appropriate "we aren't
+        #  trusted" blocked status
+        # create rbac
+        try:
+            self._create_rbac()
+        except ApiError as e:
+            # Handle forbidden error as this likely means we don't have --trust
+            if e.status.code == 403:
+                self.logger.error(
+                    "Received Forbidden (403) error from lightkube when creating required RBAC.  "
+                    "This may be due to the charm lacking permissions to create cluster-scoped "
+                    "roles and resources.  Charm must be deployed with `--trust`"
+                )
+                self.unit.status = BlockedStatus(
+                    "Cannot create required RBAC.  Charm may not have `--trust`"
+                )
+                return
+            else:
+                raise e
+
+        # Create crds
+        self._create_crds()
+
+        # deploy the controller
+        self._create_controller()
+
+        self.logger.info("Waiting for installed Kubernetes objects to be operational")
+        max_attempts = 20
+        for attempt in range(max_attempts):
+            self.logger.info(f"validation attempt {attempt}/{max_attempts}")
+            running = self._check_deployed_resources()
+
+            if running is True:
+                self.logger.info("Resources detected as running")
+                self.logger.info("Install successful")
+                self.unit.status = ActiveStatus()
+                return
+            else:
+                sleeptime = 10
+                self.logger.info(f"Sleeping {sleeptime}s")
+                time.sleep(sleeptime)
+        else:
+            self.unit.status = BlockedStatus(
+                "Some kubernetes resources missing/not ready"
+            )
+            return
+
+    def _update_status(self, event):
+        self.logger.info("Comparing current state to desired state")
+
+        running = self._check_deployed_resources()
+        if running is True:
+            self.logger.info("Resources are ok.  Unit in ActiveStatus")
+            self.unit.status = ActiveStatus()
+            return
+        else:
+            self.logger.info(
+                "Resources are missing.  Triggering install to reconcile resources"
+            )
+            self.unit.status = MaintenanceStatus(
+                "Missing kubernetes resources detected - reinstalling"
+            )
+            self._install(event)
+            return
+
+    def _remove(self, _):
+        """Remove charm"""
+        raise NotImplementedError()
+
+    def _create_crds(self, if_exists=None):
```

Comment:
My comment below about having too many helper functions applies here as well, if you see, these three functions `_create_crds()`, `_create_controller()` and `_create_rbac()` are basically doing the same. Again, just an opinion, but we can avoid this, perhaps just having one helper `_create_resource(self, rendered_obj, if_exists=None)`.

Reply:
I think this is refactored in a way you'll now like